src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs

```diff
@@ -315,7 +315,7 @@ internal virtual bool IsExplicitInterfaceImplementation
         /// Indicates whether the method is effectively readonly,
         /// by either the method or the containing type being marked readonly.
         /// </summary>
-        internal bool IsEffectivelyReadOnly => (IsDeclaredReadOnly || ContainingType?.IsReadOnly == true) && MethodKind != MethodKind.Constructor;
+        internal bool IsEffectivelyReadOnly => (IsDeclaredReadOnly || ContainingType?.IsReadOnly == true) && MethodKind != MethodKind.Constructor && !IsStatic;
```

Comment:
>!IsStatic; [](start = 149, length = 10)

should we also check that the containing type isn't a value type, as we do in `PEMethodSymbol.IsDeclaredReadOnly`?

Reply:
`PEMethodSymbol.IsDeclaredReadOnly` is meant to avoid traversing attributes in cases where `readonly` can be ruled out, in the vein of `PEMethodSymbol.IsExtensionMethod`. One thing I'm curious about is to what extent the existing concessions/optimizations around `IsExtensionMethod` are necessary for `IsReadOnly`.

It also appears that `ContainingType.IsReadOnly` returns `false` on a `readonly class` declaration due to the design of `SourceMemberContainerSymbol.MakeModifiers()`.

Reply:
Not sure I understand the `IsStatic` addition here. Shouldn't all `static` methods be implicitly readonly since they can't modify `this`?

Reply:
I looked at it the other way: "readonly" is not even a valid concept for `static` methods, so we can say they are not "readonly".
I could go either way.

Reply:
Seems like it depends on what you want the API to mean. Do you want it to mean "`this` is not writable" or do you want it to mean "`this` is a `readonly ref` parameter"?

---
In reply to: [267528694](https://github.com/dotnet/roslyn/pull/34260#discussion_r267528694) [](ancestors = 267528694)